Fix #24 Don't mess with line separators

TLmaK0 opened the pull request:
This is a try to fix #24 line separators. Works well and pass tests on my environment.

TLmaK0 commented:
I don't know how to fix travis-ci because the direct dependency with current snapshot. There aren't plugins with tests to use as template.

hseeberger commented on build.sbt:
```diff
@@ -11,3 +11,5 @@ libraryDependencies ++= List(
 
 initialCommands := """|import de.heikoseeberger.sbtheader._""".stripMargin
+
+git.useGitDescribe := true
```
Why this?

hseeberger replied:
Ah, now I see. Please add it to `project/Build.scala`.

TLmaK0 replied:
:ok_hand:

hseeberger commented on src/main/scala/de/heikoseeberger/sbtheader/HeaderPlugin.scala:
```diff
@@ -97,14 +98,16 @@ object HeaderPlugin extends AutoPlugin {
 
   private def createHeader(headerPattern: Regex, headerText: String, log: Logger)(file: File) = {
-    def write(text: String) = Files.write(file.toPath, text.split(newLine).toList, UTF_8).toFile
+    def write(text: String) = Files.write(file.toPath, text.getBytes).toFile
```
Please use `getBytes(UTF_8)`.

TLmaK0 replied:
:ok_hand:

hseeberger commented on src/main/scala/de/heikoseeberger/sbtheader/HeaderPlugin.scala:
```diff
@@ -114,4 +117,4 @@ object HeaderPlugin extends AutoPlugin {
     modifiedText.map(write)
   }
-}
+}
```
Don't remove new line.

hseeberger commented on src/main/scala/de/heikoseeberger/sbtheader/license/Apache2_0.scala:
```diff
@@ -72,4 +72,4 @@ object Apache2_0 extends License {
     }
   }
-}
+}
```
Don't remove new line.

hseeberger commented on src/main/scala/de/heikoseeberger/sbtheader/package.scala:
```diff
@@ -48,4 +48,4 @@ package object sbtheader {
 
   val newLine: String = f"%n"
-}
+}
```
Don't remove new line.

hseeberger commented on src/test/scala/de/heikoseeberger/sbtheader/HeaderPluginSpec.scala:
```diff
@@ -107,4 +107,4 @@ class HeaderPluginSpec extends WordSpec with Matchers {
     }
   }
-}
+}
```
Don't remove new line.

hseeberger commented on src/test/scala/de/heikoseeberger/sbtheader/license/Apache2_0Spec.scala:
```diff
@@ -25,24 +26,7 @@ class Apache2_0Spec extends WordSpec with Matchers {
-            | */
-            |
-            |""".stripMargin
+      val expected = Source.fromURL(getClass.getResource("/apache2")).mkString
```
Why move this?

hseeberger commented on src/test/scala/de/heikoseeberger/sbtheader/license/Apache2_0Spec.scala:
```diff
@@ -50,22 +34,7 @@ class Apache2_0Spec extends WordSpec with Matchers {
-           |# limitations under the License.
-           |
-           |""".stripMargin
+      val expected = Source.fromURL(getClass.getResource("/apache2_hash")).mkString
```
And that?

TLmaK0 replied:
only to reduce file length, I will restore the original test :)

hseeberger commented on src/main/scala/de/heikoseeberger/sbtheader/HeaderPlugin.scala:
```diff
@@ -97,14 +98,16 @@ object HeaderPlugin extends AutoPlugin {
     log.debug(s"About to create/update header for $file")
-    val (firstLine, text) = Files.readAllLines(file.toPath, UTF_8).mkString(newLine) match {
+
+    val (firstLine, text) = new Scanner(file).useDelimiter("\\Z").next() match {
```
`new Scanner(file, UTF_8)`

TLmaK0 replied:
:ok_hand:

TLmaK0 replied:
In future versions should take care about the original file encoding.

hseeberger commented:
Thanks for your contribution. Do you think this will work? I think that `Scanner` just reads the line breaks as they are and `"\n" != "\r\n"`.

TLmaK0 commented:
I think that `Scanner` read file as stream of bytes and doesn't mater which new line characters there are. `\\Z` is EOF. Copy all of then to only one string with carriage return included. Then copy all this bytes directly to final file, without change new line characters.

TLmaK0 commented:
I you use `\r\n` in your files, final file will have `\r\n`.

hseeberger commented:
I think I understand, but what about the headers? The default ones use `\n`, i.e. don't respect the original line separator. So I think you could end up with a mix of `\n` and `\r\n`, not?

TLmaK0 commented:
Well, the headers are taken from the build.sbt file, this file should have the same new line than the other source files. In my case, I work with vim in git bash for windows. System new line is '\r\n' but the editor uses '\n'. Any file I will create with this editor will have same line ending (build.sbt and src files). 

Tests will pass always, no mater if you are using autoclrf on git, when download sbt-headers src code, git will replace all line endings (or not) and you will sure you will have same line endings for all files (included build.sbt with header value).

I think there is only a posible problem: when you use autoclrf=input and then edit your build.sbt with a crlf line ending editor (or you have autoclrf=true and then use a LF line ending editor). This environment configuration it's wrong from my point of view, but anyway, the developer is able to change the build.sbt line endings and get a correct behavior. Without this patch I think isn't posible.

hseeberger commented:
"the headers are taken from the build.sbt file" => I think many use the predefined headers which always have "\n". We probably need a more general solution.

TLmaK0 commented:
sure, do you have the solution?

hseeberger commented:
Obviously not ;-)

I think we should somehow completely ignore existing line breaks for comparison and use the native line break for writing.

TLmaK0 commented:
Native line break will not work on my environment, because I use vim in a msys environment with LF endings, but scala system new line returns LFCR.
I think we should guess the line ending from the destination file, or force users to write his own license.

hseeberger commented:
I see. Guessing line endings seems a decent approach.

TLmaK0 commented on src/main/scala/de/heikoseeberger/sbtheader/HeaderPlugin.scala:
```diff
@@ -32,7 +33,7 @@ object HeaderPattern {
   def commentStartingWith(start: String): Regex =
-    new Regex(raw"""(?s)((?:$start[^\n\r]*(?:\n|\r|\r\n))+(?:\n|\r|\r\n)+)(.*)""")
+    new Regex(raw"""(?s)((?:$start[^\n\r]*(?:\n|\r|\r\n))?(?:#[^\n\r]*(?:(?:\n){2,}|(?:\r){2,}|(?:\r\n){2,})+))(.*)""")
 }
```
@hseeberger I have problems passing some tests with original regex. This new regex should fix the problem. Look at this simplified examples to see the problem:
https://regex101.com/r/sF6fB7/1 
https://regex101.com/r/rB4wD3/1

TLmaK0 commented:
@hseeberger I have added some tests. I think it should work now. I have changed Scanner by io.Source.fromInputStream due to java bug http://bugs.java.com/bugdatabase/view_bug.do?bug_id=8028387

hseeberger commented:
Pretty awesome. Thanks!

hseeberger commented:
@TLmaK0, I have published version 1.5.1-2-g8b57b53.

TLmaK0 commented:
@hseeberger thanks to you for this library :+1:
